FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -0,0 +1,137 @@
+//
+//  CoreDataFeedStore.swift
+//  FeedStoreChallenge
+//
+//  Created by Oliver Jordy Pérez Escamilla on 23/01/21.
+//  Copyright © 2021 Essential Developer. All rights reserved.
+//
+
+import Foundation
+import CoreData
+
+public final class CoreDataFeedStore: FeedStore {
+	
+	// MARK: Properties
+	lazy var context: NSManagedObjectContext = {
+		Self.coreDataStack.context
+	}()
+	private static var coreDataStack: CoreDataStack = CoreDataStack(modelName: "FeedStoreModel")
+	
+	// MARK: Initializers
+	public init(coreDataStack: CoreDataStack) {
+		Self.coreDataStack = coreDataStack
+	}
+	
+	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
+		context.perform { [unowned context] in
+			
+			do {
+				guard let cache = try self.fetchCache() else { return completion(nil) }
+				context.delete(cache)
+				
+				completion(nil)
+				
+			} catch {
+				return completion(error)
+			}
+			
+			Self.coreDataStack.saveContext()
+		}
+	}
+	
+	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
+		context.perform { [unowned self] in
+			
+			do {
+				if let cache = try self.fetchCache() {
+					self.context.delete(cache)
+				}
+				
+			} catch {
+				return completion(error)
+			}
+			
+			let feedCache = FeedCache(context: self.context)
+			let feedImages = transformToManagedRepresentation(feed)
+			
+			feedCache.feedImages = NSOrderedSet(array: feedImages)
+			feedCache.timestamp = timestamp
+			
+			do {
+				try self.context.save()
+			} catch {
+				completion(error)
+			}
+			completion(nil)
```

Comment:
The `completion` closure should only be called once, with either an `error` or `nil`. But this method can call the `completion` closure more than once.

To solve it and simplify the logic, the insert operation should all be within the do/catch block, since the `save` operation can fail:

```suggestion
			do {
				if let cache = try self.fetchCache() {
					self.context.delete(cache)
				}
				
				let feedCache = FeedCache(context: self.context)
				let feedImages = transformToManagedRepresentation(feed)
				
				feedCache.feedImages = NSOrderedSet(array: feedImages)
				feedCache.timestamp = timestamp
				try self.context.save()
				completion(nil)
			} catch {
				completion(error)
			}
```